src/Interpreters/parseColumnsListForTableFunction.cpp

```diff
@@ -71,6 +73,29 @@ void validateDataType(const DataTypePtr & type, const DataTypeValidationSettings
                 "Set setting allow_experimental_variant_type = 1 in order to allow it", type->getName());
         }
     }
+
+    if (const auto * nullable_type = typeid_cast<const DataTypeNullable *>(type.get()))
+    {
+        validateDataType(nullable_type->getNestedType(), settings);
+    }
+    else if (const auto * lc_type = typeid_cast<const DataTypeLowCardinality *>(type.get()))
+    {
+        validateDataType(lc_type->getDictionaryType(), settings);
+    }
+    else if (const auto * array_type = typeid_cast<const DataTypeArray *>(type.get()))
+    {
+        validateDataType(array_type->getNestedType(), settings);
+    }
+    else if (const auto * tuple_type = typeid_cast<const DataTypeTuple *>(type.get()))
+    {
+        for (const auto & element : tuple_type->getElements())
+            validateDataType(element, settings);
+    }
```

Comment:
Shouldn't the same thing be done for variant also?

Reply:
Good catch! Yes, we should validate variants also, will fix